### DISCUSSION TITLE
Suggestion: How to use some reflection for `Registry.cs`?

### DISCUSSION THREAD

**Discussion by riemannulus** [planetarium/libplanet#2288] [Ideas]
## Context
The problem with the current `Registry.cs` is to create a constructor that creates an empty object to deserialize an action.

This is problematic for Actions that use non-nullable fields.

1. Entering an arbitrary value is likely to cause unintended behavior.
2. Changing fields to nullable causes increase code complexity or undefined behavior.

The cause of this behavior is that SysAction reflection is implemented by directly creating an empty object during the instantiation process and putting the value into `LoadPlainValue`.
https://github.com/planetarium/libplanet/blob/8ff56770f28f8fcb808370a417b61c7aa7b28fc3/Libplanet/Action/Sys/Registry.cs#L48-L49

## Idea
If we use [`Activator.CreateInstance`](https://docs.microsoft.com/en-us/dotnet/api/system.activator.createinstance?view=net-6.0), maybe solve this problem.
I wrote some code for a basic idea. (not tested)
https://gist.github.com/riemannulus/7405e0d361364c6afa0ab433905ae81c

Have any opinion on this issue? @libplanet

**Comment by riemannulus** [comment]
AFAIK, @dahlia wrote this code. Could you have some advice on this suggestion?

**Comment by longfin** [comment]
- `Activator.CreateInstance()` had been discouraged due to [IL2CPP](https://docs.unity3d.com/2021.3/Documentation/Manual/IL2CPP.html) compatibility. (we may need to investigate furthermore)
- Since system actions has been treated as `IAction`, there still seems to be code somewhere to initialize it with a parameter-less constructor. thus, it may be a story about `IAction` (de)serialization, not just `Registry.cs`.

**Comment by riemannulus** [comment]
`ActionEvaluator` 와 `TypeLoader` 의 리팩토링으로 인해 해당 안은 더 이상 유효하지 않으므로 닫겠습니다.
